mono/metadata/image.c

```diff
@@ -2133,41 +2311,34 @@ mono_image_close_except_pools (MonoImage *image)
 		}
 	}
 
-#ifdef HOST_WIN32
-	mono_images_lock ();
-	if (image->is_module_handle && !image->has_entry_point)
-		FreeLibrary ((HMODULE) image->raw_data);
-	mono_images_unlock ();
-#endif
+	/* a MonoDynamicImage doesn't have any storage */
+	g_assert (image_is_dynamic (image) || image->storage != NULL);
 
-	if (image->raw_buffer_used) {
-		if (image->raw_data != NULL) {
-#ifndef HOST_WIN32
-			if (image->fileio_used)
-				mono_file_unmap_fileio (image->raw_data, image->raw_data_handle);
-			else
-#endif
-				mono_file_unmap (image->raw_data, image->raw_data_handle);
-		}
-	}
-	
-	if (image->raw_data_allocated) {
+	if (image->storage && m_image_is_raw_data_allocated (image)) {
 		/* FIXME: do we need this? (image is disposed anyway) */
-		/* image->raw_metadata and cli_sections might lie inside image->raw_data */
+		/* image->raw_metadata and cli_sections might lie inside m_image_get_raw_data (image) */
 		MonoCLIImageInfo *ii = image->image_info;
 
-		if ((image->raw_metadata > image->raw_data) &&
-			(image->raw_metadata <= (image->raw_data + image->raw_data_len)))
+		if ((image->raw_metadata > m_image_get_raw_data (image)) &&
+			(image->raw_metadata <= (m_image_get_raw_data (image) + m_image_get_raw_data_len (image))))
 			image->raw_metadata = NULL;
 
 		for (i = 0; i < ii->cli_section_count; i++)
-			if (((char*)(ii->cli_sections [i]) > image->raw_data) &&
-				((char*)(ii->cli_sections [i]) <= ((char*)image->raw_data + image->raw_data_len)))
+			if (((char*)(ii->cli_sections [i]) > m_image_get_raw_data (image)) &&
+				((char*)(ii->cli_sections [i]) <= ((char*)m_image_get_raw_data (image) + m_image_get_raw_data_len (image))))
 				ii->cli_sections [i] = NULL;
 
-		g_free (image->raw_data);
 	}
 
+#ifdef HOST_WIN32
+	mono_images_lock ();
+	if (m_image_is_module_handle (image) && !m_image_has_entry_point (image))
+		FreeLibrary ((HMODULE) m_image_get_raw_data (image));
```

Comment:
Hm. Pretty sure I need to move this to the `MonoImageStorage` destructor.  Don't want to  call `FreeLibrary` if it's still shared with another image.